Laromance model in NEML2

lynnmunday opened the pull request:
closes #29774

lynnmunday committed:
laromance grade 91 from type model implemented in NEML2.
This is based on a grade 91 model from LANL on 2/1/2025.
The implementation uses a dummy inteprolation grid, values, and transforms and does not contain any real material data.
The actual grade 91 data is in Bison.
This implements multilinear interpolation.
There are 6 interpolation axes and 3 values at each node.
The interpolation axes and interpolated values are transformed.
This contains examples where the model only performs interpolation and a full example with radial return done in NEML2.
The radial return mapping requries a change to NEML2 in math.cxx that Gary is including in his neml2 udpate
closes #29774

lynnmunday commented on modules/solid_mechanics/test/tests/neml2/interpolated_material/tests:
This is not set up at all.

lynnmunday commented on modules/solid_mechanics/test/tests/neml2/interpolated_material/moose_neml2_return_mapping.i:
```diff
@@ -0,0 +1,183 @@
+    verbose = true
+    device = 'cpu'
+
+    moose_input_types = 'MATERIAL MATERIAL MATERIAL MATERIAL MATERIAL MATERIAL MATERIAL POSTPROCESSOR POSTPROCESSOR'
```
This is not set up correctly because I am getting the error:
```The NEML2 model requires a stateful input variable `old_state/cell_dd`, but its state counterpart on the output axis has not been retrieved by any object. Therefore, there is no way to properly propagate the corresponding stateful data in time. The common solution to this problem is to add a NEML2ToMOOSE retriever such as those called `NEML2To*MOOSEMaterialProperty`.```

hugary1995 replied:
I think you need to save the corresponding neml2 output variable into moose. The output variable should be `state/cell_dd`.

lynnmunday replied:
Gary gave me a fix for the batch size problem I had.  The fix goes in NEML2.  It is from linear algebra library changing between libtorch versions.  I am using pytorch version 2.1.2.post4 and I need this fix.  Gary is using 2.5.1 and does not need the fix.

lynnmunday commented:
@dschwen @hugary1995 This is the LAROMANCE implementation in NEML2.  I am still converting the radial return mapping input file over to the new syntax.  I left a comment about the error I'm getting.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2662105/), step Docs: sync website on 9474c7d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29775/site)

This comment will be updated on new commits.

lynnmunday committed:
trying to refactor into a single class that reads all data from json but getting slightly different results

lynnmunday committed:
Using a json file created with a seed=0 in the python script.
WE can now create add the transforms to the json files.
I also changed the interpolated material gold file to use the same transforms as the radial return mapping.

lynnmunday committed:
changed the multilinear interpolation material to read output transforms from json file which means I no longer need derived classes for output strain rate and dislocation densities.

lynnmunday committed:
using enums for input grid transforms too

lynnmunday committed:
adding exceptions instead of asserts

moosebuild commented:
Job [Conda build (ARM Mac)](https://civet.inl.gov/job/2686668/) on efce148 : invalidated by @lynnmunday

lynnmunday committed:
renamed neml2 interpolation objects to laromance6Dinterpolation
Added documentation

lynnmunday committed:
fixing requirement in tests file

lynnmunday commented on modules/solid_mechanics/src/neml2/userobjects/NEML2ModelExecutor.C:
```diff
@@ -269,6 +269,8 @@ NEML2ModelExecutor::applyPredictor()
 {
   if (!model().input_axis().has_state())
     return;
+  if (!model().input_axis().has_old_state())
```
Gary might have pushed this in his pr.

lynnmunday commented on modules/solid_mechanics/test/tests/neml2/laromance/models/create_6D_dummy_grid_json.py:
Can I include this file for making the dummy json interpolation grid?

lynnmunday committed:
removing fixme comments

lynnmunday commented:
@dschwen will you review and merge?

lynnmunday commented:
@dschwen will you review?

hugary1995 commented on modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LAROMANCE6DInterpolation.md:
```diff
@@ -0,0 +1,14 @@
+# LAROMANCE6DInterpolation
+
+!alert note
+This is a custom NEML2 object for solid_mechanics. To enable it, configure MOOSE with libtorch and initialize the NEML2 submodule.
```
Maybe link to the NEML2 doc?

hugary1995 commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,374 @@
+  options.set<std::string>("model_file_name");
+  options.set<std::string>("model_file_variable_name");
+
+  options.set<bool>("_use_AD_first_derivative") = true;
+  options.set<bool>("_use_AD_second_derivative") = true;
```
```suggestion
```
The API for requesting AD has changed. Take a look at `NEML2TestModel.cxx`

lynnmunday replied:
@dschwen helped me get this to work but I had to turn off jit.

hugary1995 reviewed (changes requested):
Minor comments. The NEML2 part otherwise looks good to me. Did you try rerunning the models after the NEML2 submodule update? The submodule update enables JIT for all NEML2 models by default.

lynnmunday committed:
updating to new AD interface.  This model does not work with jit so that had to be disabled.

lynnmunday commented:
@hugary1995 I addressed your comments.  @dschwen Will you review and merge?

dschwen commented on modules/solid_mechanics/neml2/include/LAROMANCE6DInterpolation.h:
```diff
@@ -0,0 +1,128 @@
+#pragma once
```
Missing header

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,385 @@
+  // JSON
+  options.set<std::string>("model_file_name");
+  options.set<std::string>("model_file_variable_name");
+  // jit doe snot currently work with this
+  options.set<bool>("jit") = false; // false;
```
```suggestion
  // jit does not currently work with this
  options.set<bool>("jit") = false;
```

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,385 @@
+{
+  switch (transform_type)
+  {
+    case TransformEnum::COMPRESS:
+    {
+      Scalar transformed_result = transform_compress(data, param);
+      return transformed_result;
+    }
+    case TransformEnum::DECOMPRESS:
+    {
+      Scalar transformed_result = transform_decompress(data, param);
+      return transformed_result;
+    }
```
```suggestion
    case TransformEnum::COMPRESS:
      return transform_compress(data, param);

    case TransformEnum::DECOMPRESS:
      return transform_decompress(data, param);
```

etc

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,385 @@
+            for (auto && value : level4)
+              linearize_values.push_back(value);
+
+  long long sz_l0 = out_data.size();
+  long long sz_l1 = out_data[0].size();
+  long long sz_l2 = out_data[0][0].size();
+  long long sz_l3 = out_data[0][0][0].size();
+  long long sz_l4 = out_data[0][0][0][0].size();
+  long long sz_l5 = out_data[0][0][0][0][0].size();
```
```suggestion
  const auto sz_l0 = out_data.size();
  const auto sz_l1 = out_data[0].size();
  const auto sz_l2 = out_data[0][0].size();
  const auto sz_l3 = out_data[0][0][0].size();
  const auto sz_l4 = out_data[0][0][0][0].size();
  const auto sz_l5 = out_data[0][0][0][0][0].size();
```

there doesn't seem to be any checking that _all_ items in a dimension have the same size. Furthermore...

dschwen replied:
move this stuff above teh linearization and at each level add a check to ensure that each item has the same size at the 0 item

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,385 @@
+  long long sz_l3 = out_data[0][0][0].size();
+  long long sz_l4 = out_data[0][0][0][0].size();
+  long long sz_l5 = out_data[0][0][0][0][0].size();
+  return Scalar(
+      torch::from_blob(linearize_values.data(), {sz_l0, sz_l1, sz_l2, sz_l3, sz_l4, sz_l5})
+          .clone());
```
I'm surprised this doesn't utterly fail for you. `from_blob` requires a pointer to a contiguous block of values. You are setting this up as nested vectors, which contain a control block and have data allocated on the heap. There is _zero_ guarantee that this results in contiguous memory.

dschwen reviewed (changes requested):
<s>from_blob needs fixing</s> I misread this. But lets check the shape more thoroughly

lynnmunday committed:
Addressing reviewer comments and added test to check error

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2737112/) on 426abae : invalidated by @lynnmunday

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2737115/), step Verify coverage on 426abae wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lynnmunday commented:
@dschwen I addressed your comments, Will you review and merge?

dschwen commented on modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LAROMANCE6DInterpolation.md:
```diff
@@ -0,0 +1,14 @@
+!alert note
+This is a custom [NEML2](solid_mechanics/NEML2.md) object for solid_mechanics. To enable it, configure MOOSE with libtorch and initialize the [NEML2 submodule](https://github.com/applied-material-modeling/neml2).
+
+The documentation is for a NEML2 [!citep](neml2osti, neml2_anl_report) implementation of the LAROMance surrogate model.  The LAROMance surrogate model computes the inelastic effective strain rate used in an isotropic viscoplastic constitutive model update [!cite](tallman-2020).  The LAROMance surrogate model is fit to a database of polycrystalline crystal plasticity simulations.  The material's m ...
```
```suggestion
The documentation is for a NEML2 [!citep](neml2osti, neml2_anl_report) implementation of the LAROMance surrogate model.  The LAROMance surrogate model computes the inelastic effective strain rate used in an isotropic viscoplastic constitutive model update [!cite](tallman-2020).  The LAROMance surrogate model is fit to a database of polycrystalline crystal plasticity simulations.  The material's microstructure, as described by the distribution of dislocation content, temperature, stress, and environment factor (precipitate phase fraction, neutron fluence, etc. ), varied in the crystal plasticity creep simulations used to create the database.  The inelastic strain history and mobile and immobile dislocation evolution from the crystal plasticity simulations are then fit to an element-based LAROMance model [!citep](ruybalid_osti24, munday_osti25). The element-based input space provides an interpolation grid for the output state of the material.  This version of the LAROMance model uses piecewise-continuous linear functions to interpolate the input grid.  The six-dimensional (6D) input space for the LAROMance model, $f\left(T,\epsilon_{p},\sigma_{vm},\rho_c,\rho_w,\Phi_{mx}\right)$, corresponds to the variables varied in the LApx simulations---that is, temperature ($T$ in units K), von Mises stress ($\sigma_{vm}$ in units MPa), accumulated effective inelastic strain ($\epsilon_{p}$), mobile or cell dislocation density ($\rho_c$ in units $m^{-2}$), immobile or wall dislocation density ($\rho_w$ in units $m^{-2}$), and environmental factor ($\Phi_{mx}$ in units of number density; $count/m^{-3}$).  Interpolated outputs from the LAROMance model are inelastic effective strain rate ($\dot{\epsilon}_{p}$), and mobile and immobile dislocation density evolution rates ( $\dot{\rho}_c$ and  $\dot{\rho}_w$ ).  Transforms are applied to the input grid and output state to normalize the interpolation space, apply constraints on the output, and allow linear regression to fit a nonlinear relationship between inputs and outputs such as the logarithmic evolution of dislocation density due to accumulated effective inelastic strain [!cite](tallman-2020).  Additional information about the LAROMance models and a previous implementation is available [here](/LAROMANCE.md).
```

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,409 @@
+LAROMANCE6DInterpolation::LAROMANCE6DInterpolation(const OptionSet & options)
+  : Model(options),
+    _s(declare_input_variable<Scalar>("von_mises_stress")),
+    _T(declare_input_variable<Scalar>("temperature")),
```
I'd refrain from using variables consisting of a single capital letter with an underscore prefix. According to the C++ standard, those are reserved for internal purposes, and we have run into issues with this before.

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,409 @@
+}
+
+LAROMANCE6DInterpolation::TransformEnum
+LAROMANCE6DInterpolation::get_transform_enum(const std::string & name) const
```
this sure is easier with `MooseEnum` and `getEnum`

lynnmunday replied:
I can't use those, right?

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,409 @@
+          for (const auto m : {0, 1})
+            for (const auto n : {0, 1})
+            {
+
```
```suggestion
```

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,409 @@
+  Scalar right_coord = grid.index({left_idx + torch::tensor(1, default_integer_tensor_options())});
+  Scalar left_fraction = (right_coord - interp_points) / (right_coord - left_coord);
+
+  return std::pair<Scalar, Scalar>(left_idx, torch::stack({left_fraction, 1 - left_fraction}, -1));
```
```suggestion
  return {left_idx, torch::stack({left_fraction, 1 - left_fraction}, -1)};
```

dschwen commented on modules/solid_mechanics/neml2/src/LAROMANCE6DInterpolation.cxx:
```diff
@@ -0,0 +1,409 @@
+
+/// compute interpolated value
+Scalar
+LAROMANCE6DInterpolation::interpolate_and_transform()
```
all variables here should be `const`

dschwen commented on modules/solid_mechanics/solid_mechanics.mk:
```diff
@@ -1,5 +1,6 @@
 # NEML2
 ADDITIONAL_NEML2_DIRS += $(APPLICATION_DIR)/test/neml2
+ADDITIONAL_NEML2_DIRS += $(CURDIR)/neml2
```
why did you choose `$(CURDIR)` here? I'm not sure this is correct (when using this module from another app).

dschwen commented on modules/solid_mechanics/test/tests/neml2/laromance/laromance_interpolate.i:
```diff
@@ -0,0 +1,126 @@
+    type = GeneratedMeshGenerator
+    dim = 1
+  []
+[]
+[Variables]
+  [dummy_var]
+  []
+[]
+[Problem]
+  kernel_coverage_check = false
+[]
```
```suggestion
[]

[Variables]
  [dummy_var]
  []
[]

[Problem]
  kernel_coverage_check = false
[]
```

dschwen commented on modules/solid_mechanics/test/tests/neml2/laromance/laromance_return_mapping.i:
```diff
@@ -0,0 +1,158 @@
+    type = ParsedFunction
+    expression = 200 #MPa
+  []
+[]
+[Materials]
```
```suggestion
[]

[Materials]
```

dschwen commented on modules/solid_mechanics/test/tests/neml2/laromance/models/create_6D_dummy_grid_json.py:
```diff
@@ -0,0 +1,94 @@
+import json
```
Let's put a header here as well

dschwen reviewed (changes requested):
See comments

lynnmunday committed:
Address review comments.
Added better documentation for the LAROMANCE NEML2 material model input.
Made more functions const
Removed default variables NEML2 OptionSet
Renamed variables in NEML2 class

dschwen commented on modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LAROMANCE6DInterpolation.md:
```diff
@@ -0,0 +1,18 @@
+!alert note
+This is a custom [NEML2](solid_mechanics/NEML2.md) object for solid_mechanics. To enable it, configure MOOSE with libtorch and initialize the [NEML2 submodule](https://github.com/applied-material-modeling/neml2).
+
+The documentation is for a NEML2 [!citep](neml2osti, neml2_anl_report) implementation of the LAROMance surrogate model.  The LAROMance surrogate model computes the inelastic effective strain rate used in an isotropic viscoplastic constitutive model update [!cite](tallman-2020).  The LAROMance surrogate model is fit to a database of polycrystalline crystal plasticity simulations.  The material's m ...
```
"and and", "avaiable"

lynnmunday replied:
good catch!

lynnmunday committed:
Adding reviewer comments that were acciently overwritten

lynnmunday commented:
@dschwen Will you review and merge

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2756751/) on 35aa387 : invalidated by @lynnmunday

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/2756720/) on 35aa387 : invalidated by @lynnmunday

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2756766/), step Verify coverage on 35aa387 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2756766/), step Verify coverage on 35aa387 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2794494/) on 36ea0ac : invalidated by @lynnmunday

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2794494/) on 36ea0ac : invalidated by @lynnmunday

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2794448/) on 36ea0ac : invalidated by @lynnmunday

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2794494/) on 36ea0ac : invalidated by @lynnmunday

lynnmunday commented:
A different random parallel restart test fails each time I invalidate the OpenMPI tests.  These failures are not related to this commit. @dschwen can you merge it anyway?

hugary1995 commented:
it's passing now

lynnmunday commented:
@dschwen Now its passing, will you merge?
